flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestServerHandler.java

```diff
@@ -82,16 +82,17 @@ protected void channelRead0(ChannelHandlerContext ctx, NettyMessage msg) throws
 
                 LOG.debug("Read channel on {}: {}.", ctx.channel().localAddress(), request);
 
+                // Always register reader before requesting the subpartition in order to
+                // subsequent requests will be sure that PartitionRequest was received already
+                // even if subpartition view has not created yet.
+                NetworkSequenceViewReader reader =
+                        outboundQueue.notifyReaderCreated(
+                                new CreditBasedSequenceNumberingViewReader(
+                                        request.receiverId, request.credit, outboundQueue));
```

Comment:
Won't this cause resource leaks? What if `PartitionRequest` never succeeds and `allReaders.remove(toCancel);` in `org.apache.flink.runtime.io.network.netty.PartitionRequestQueue#userEventTriggered` never happens?

Reply:
Why is it a leak? If it never succeeds we just will be holding one reader for one subpartition until restart happens, so no leaks here.
Just in case, notifyReaderCreated doesn't add a reader every time, it just holds the first one and ignores the others.

Reply:
> until restart happens

I'm not sure, but it might be until TM restarts, which might be never, while number of those zombie readers would be increasing with the number of scheduled jobs in the cluster. Think about a use case where there is a single cluster and many adhoc short lived jobs.